File: medium/56_merge_intervals.rb
# 56. Merge Intervals
# Medium
#
# Given an array of intervals where intervals[i] = [starti, endi], merge all
# overlapping intervals, and return an array of the non-overlapping intervals
# that cover all the intervals in the input.

def merge(intervals)
  sorted = intervals.sort { |x, y| x[0] <=> y[0] }

  res = []
  cur = sorted[0]

  (1...sorted.size).each do |i|
    if cur[1] >= sorted[i][0]
      cur = [cur[0], [cur[1], sorted[i][1]].max]
    else
      res << cur
      cur = sorted[i]
    end
  end

  res << cur
  res
end

p merge([[1, 3], [2, 6], [8, 10], [15, 18]]) # [[1,6],[8,10],[15,18]]
p merge([[1, 4], [4, 5]])                    # [[1,5]]

File: python/medium/56_merge_intervals.py
"""
56. Merge Intervals
Medium

Given an array of intervals where intervals[i] = [starti, endi], merge all
overlapping intervals, and return an array of the non-overlapping intervals
that cover all the intervals in the input.
"""


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])

        res = []
        prev = intervals[0]

        for i in range(1, len(intervals)):
            if prev[1] >= intervals[i][0]:
                prev = [prev[0], max(prev[1], intervals[i][1])]
            else:
                res.append(prev)
                prev = intervals[i]

        res.append(prev)

        return res
